---
title: "Differential gene expression across organoid phenotypes"
author: "Benedikt Rauscher"
date: "11/15/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, message = F, warning = F, results = 'hide')
```

# Dependencies

```{r}
library(limma)
library(clusterProfiler)
library(ReactomePA)
library(enrichplot)
library(org.Hs.eg.db)
library(pheatmap)
library(reshape2)
library(tidyverse)
library(cowplot)
library(here)
```

Define function namespace.

```{r}
filter <- dplyr::filter
select <- dplyr::select
rename <- dplyr::rename
slice <- dplyr::slice
```

Set parameters for plotting.

```{r}
theme_set(theme_cowplot())
```

# Data loading

We load the gene expression data from an Rdata object from file.

```{r}
load(here('../../data/gene_expression/promise_expr.rda'))

## annotate phenotype group
solid <- c('D004', 'D007', 'D010', 'D019', 'D020', 
           'D022', 'D046', 'D054', 'D055')
cystic <- c('D013', 'D018', 'D021', 'D027', 'D030')

## exclude lines 15 and 52 due to potential contamination
promise_expr <- promise_expr %>% 
  mutate(phenotype = ifelse(line %in% solid, 'solid', 
                     ifelse(line %in% cystic, 'cystic', 'other'))) %>% 
  filter(phenotype != 'other')
```

# Quality control

We perform a clustering to check if the samples are grouped correctly.

```{r}
## make expr matrix
expr_mat <- promise_expr %>% 
  distinct(probe, line, date, expr) %>%
  acast(probe ~ line + date, value.var = 'expr')

## generate heatmap annotation
anno <- promise_expr %>% 
  distinct(line, date, batch_large, phenotype) %>% 
  unite(id, line, date) %>% 
  mutate_all(as.factor) %>%
  as.data.frame() %>% column_to_rownames('id')

## visualize distance matrix
pheatmap(as.matrix(dist(t(expr_mat))),
         annotation_row = anno)

## visualize most variable probes
pheatmap(expr_mat[order(apply(expr_mat, 1, sd), decreasing = T)[1:500],], 
         annotation_col = anno)
```

'D021_2017-05-19' looks like a technical outlier so we exclude it from the data. We do not see a clear separation of cystic and solid organoid lines based on the clustering.

```{r}
## exclude outlier
promise_expr <- promise_expr %>% filter(!(line == 'D021' & date == '2017-05-19'))
```

# Differential gene expression

We use 'limma' to perform differential gene expression analysis between cystic and solid organoid lines.

```{r}
## make new expr matrix
expr_mat <- promise_expr %>% 
  distinct(probe, line, date, expr) %>%
  acast(probe ~ line + date, value.var = 'expr')

## sample annotation
sample_anno <- promise_expr %>% 
  distinct(line, date, phenotype) %>% 
  unite(id, line, date, remove = F) %>%
  arrange(line)

## model matrix
mm <- model.matrix(~phenotype, data = sample_anno)

## set technical replicate plots
blocks <- as.factor(sample_anno$line)
## compute within-block correlations
corfit <- duplicateCorrelation(expr_mat, mm, block = blocks)

## fit model and compute model coefficients
fit_phenotype <- lmFit(expr_mat, mm, 
                       block = blocks, 
                       cor= corfit$consensus.correlation)
fit_phenotype <- eBayes(fit_phenotype)

## extract results and annotate gene symbols
probe_to_symbol <- promise_expr %>% distinct(probe, symbol) %>% 
  group_by(probe) %>% slice(1) %>% ungroup()
phenotype_res <- topTable(fit_phenotype, coef = 'phenotypesolid', n = Inf) %>% 
  as_tibble(rownames = 'probe') %>%
  left_join(probe_to_symbol)
```

We plot a histogram of the P-values to check if we have good signal.

```{r}
phenotype_res %>% ggplot(aes(P.Value)) + geom_histogram(bins = 50)
```

We also plot the top 3-4 differentially expressed genes to check expression differences.

```{r}
promise_expr %>% filter(probe %in% phenotype_res$probe[1:4]) %>% 
  ggplot(aes(phenotype, expr)) +
  stat_summary(fun = 'mean', color = '#cccccc', geom = 'bar') + 
  geom_point(aes(color = line)) + 
  facet_wrap(~symbol + probe)
```

# Gene set enrichment

## Gene Ontology

I use the clusterProfiler + fgsea packages to perform a ranked gene set enrichment analysis in order to identify differentially expressed pathways.

```{r}
## annotate entrez gene ids
probe_to_entrez <- promise_expr %>% distinct(probe, entrez) %>% 
  group_by(probe) %>% slice(1) %>% ungroup()

## select strongest probe for each entrez id
res_filtered <- phenotype_res %>% inner_join(probe_to_entrez) %>% 
  group_by(entrez) %>% top_n(1, -P.Value) %>% ungroup()

## create ranked list for enrichment
ranks <- setNames(res_filtered$t, as.character(res_filtered$entrez))
ranks <- sort(ranks, decreasing = T)

## GO term enrichment (biological process)
gse_go <- gseGO(
  geneList = ranks,
  OrgDb = org.Hs.eg.db,
  ont = 'BP',
  nPerm = 1e5,
  minGSSize = 100,
  maxGSSize = 500
)
```

We can generate a network plot using clusterProfiler in order to get an overview of the results.

```{r}
emapplot(gse_go)
```

## Reactome

We do the same for Reactome instead of GO terms. A convenience function for enrichment against Reactome is implemented in the 'ReactomePA' package. Conceptually this function works the same way as the clusterProfiler functions.

```{r}
## Reactome enrichment analysis
gse_reactome <- gsePathway(
  geneList = ranks,
  organism = 'human',
  nPerm = 1e5,
  minGSSize = 100,
  maxGSSize = 500
)
```

Again we plot an enrichment map to visualize the results.

```{r}
emapplot(gse_reactome)
```

## Intestinal crypt signatures

We perform GSEA for the humanized intestinal crypt signatures published in Merloz-Suarez et al (Cell Stem Cell, 2011).

```{r}
## parse signatures and make list
intestinal_sig <- readxl::read_excel(here('../../data/gene_expression/merloz-suarez_sigantures.xls'), 
                                     sheet = 1, skip = 4) %>% .[,1:4] %>%
  gather(signature, symbol) %>% drop_na() %>% 
  mutate(symbol = gsub('\\*', '', symbol))

## run gsea with clusterprofiler
res_symbol <- phenotype_res %>% 
  group_by(symbol) %>% top_n(1, -P.Value) %>% ungroup()
ranks_symbol <- sort(setNames(res_symbol$t, res_symbol$symbol), decreasing = T)

gse_sig <- GSEA(
  geneList = ranks_symbol,
  TERM2GENE = intestinal_sig,
  nPerm = 1e5,
  minGSSize = 1,
  maxGSSize = 1000,
  pvalueCutoff = 1
)

## output as tibble
gse_sig_tbl <- as_tibble(gse_sig)
```

We see a significant upregulation of the proliferation signature and a downregulation of the Lgr5 ISC signature in solid compared to cystic organoids. We generate barcode plots to visualize these enrichments.

```{r}
## lgr5 signature
gseaplot2(
  gse_sig, geneSetID = gse_sig$ID[2], 
  title = paste0(gse_sig$ID[2], 
                ' (p = ', round(gse_sig_tbl$pvalue[2], 3), 
                '; NES = ', round(gse_sig_tbl$NES[2], 2), ')')
) 

## proliferation
gseaplot2(
  gse_sig, geneSetID = gse_sig$ID[1],
  title = paste0(gse_sig$ID[1], 
                ' (p = ', round(gse_sig_tbl$pvalue[1], 3), 
                '; NES = ', round(gse_sig_tbl$NES[1], 2), ')')
)
```

# Session info

```{r, results = 'show'}
sessionInfo()
```

