---
title: "Tublin inhibitors in colorectal cancer PDOs"
author: "Benedikt Rauscher"
date: "7/1/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Dependencies

```{r, results='hide', warning=F, message=F}
library(pheatmap)
library(openxlsx)
library(plotly)
library(cowplot)
library(fgsea)
library(affy)
library(clusterProfiler)
library(limma)
library(reshape2)
library(tidyverse)
```

```{r}
theme_set(theme_cowplot())
```

# Analysis

Idea: there is data indicating that tubulin inhibitors might affect Wnt signaling, because tubulin is required to transport CTNNB1 into the nucleus. In the context of the PDO study we want to look into this.

## Viability response of tubulin inhibition in PDOs

We have 3 tubulin inhibitors in our clinical library (Paclitaxel, Docetaxel, Vinblastin). I want to first check whether these inhibitors display any viability phenotypes in PDOs and whether there is a differential effect between different lines. First I load CTG for all PDO lines including lines not previously described in the manuscript / preprint. I load the data from an Rdata file, prepared according to the markdown 'ctg_analysis_pancancer.Rmd'.

```{r, results='hide', warning=F, message=F,eval=F}
## load data from Rdata
load('../data/pdo_ctg_19.rda')

## negative viability make no sense and are set to 0
pdo_ctg_19 <- pdo_ctg_19 %>% mutate(viability = ifelse(viability < 0, 0, viability))
```

This dataset consists of screens performed using 2 different libraries (one is an 'evolved' version of the other). Therefore, I cluster normalized viability phenotypes for drugs that overlap between two libraries. I want to rule out that there are batch effects that might confound analyses, where drug effects are compared between lines that have been screened with either library 08 or library 11.

```{r, results='hide', warning=F, message=F,eval=F}
## shared drugs between libraries 8 and 11
shared_drugs <- pdo_ctg_19 %>% distinct(library, drug) %>% 
  mutate(has = T) %>% spread(library, has) %>% 
  filter(`08` & `11`) %>% pull(drug)

## controls are also excluded
controls <- c('DMSO', 'Staurosporine_500nM', 'Staurosporine')

## annotationi of library
library_anno <- pdo_ctg_19 %>% 
  distinct(unique_id, library, tissue_of_origin, sample_type) %>% 
  as.data.frame() %>% column_to_rownames('unique_id')

## cluster using heatmap
pdo_ctg_19 %>% 
  filter(!drug %in% controls,
         drug %in% shared_drugs) %>% 
  acast(drug+rack.concentration ~ unique_id, value.var = 'viability') %>%
  pheatmap(annotation_col = library_anno)
```

It does seem to me that there could be differences here. I use a statistical test to identify drug-concentration pairs that significantly differ between the two libraries.

```{r, results='hide', warning=F, message=F,eval=F}
## perform test
lib_diff <- pdo_ctg_19 %>% filter(drug %in% shared_drugs, !drug %in% controls) %>% 
  group_by(drug, rack.concentration) %>% 
  group_modify(~ broom::tidy(t.test(viability ~ library, data = .x))) %>% 
  ungroup() %>% arrange(p.value) %>% 
  mutate(FDR = p.adjust(p.value, method='BH'))

## how many are significantly different?
(lib_diff %>% filter(FDR < 0.05) %>% nrow())/nrow(lib_diff)
```

45% of all drug-concentration pairs are significantly different. I will look at some plots to look at those individual differences but the likely conclusion is that the two libraries should not be compared at all.

```{r, results='hide', warning=F, message=F,eval=F}
## venetoclax (0.008) is a strong hit 
pdo_ctg_19 %>% filter(drug == 'Venetoclax') %>% 
  ggplot(aes(rack.concentration, viability, colour = library, group = unique_id)) + 
  geom_point() + geom_line() + 
  ggtitle('Venetoclax')

## taxanes
pdo_ctg_19 %>% filter(drug %in% c('Paclitaxel', 'Docetaxel')) %>% 
  ggplot(aes(rack.concentration, viability, colour = library, group = unique_id)) + 
  geom_point() + geom_line() + facet_wrap(~drug) +
  ggtitle('Taxane comparison between libraries 8 and 11')

## MEK inhibitors
pdo_ctg_19 %>% filter(drug %in% c('Trametinib', 'Binimetinib')) %>% 
  ggplot(aes(rack.concentration, viability, colour = library, group = unique_id)) + 
  geom_point() + geom_line() + facet_wrap(~drug) +
  ggtitle('MEK inhibitor comparison between libraries 8 and 11')
```

I think for any clustering/heatmap visualizations the two libraries have to be considered separately. I now calculate area und the dose-response-curve statistics.

```{r, results='hide', warning=F, message=F,eval=F}
aucs <- pdo_ctg_19 %>% filter(! drug %in% c('DMSO', 'Staurosporine_500nM', 'Staurosporine')) %>% 
  filter(!((drug == 'Bortezomib') & (rack.concentration == '0.04'))) %>%
  dplyr::select(donor, tissue_of_origin, sample_type, library,
                drug, rack.concentration, viability) %>% 
  mutate(viability = ifelse(viability < 0, 0, ifelse(viability > 1, 1, viability))) %>%
  group_by(donor, tissue_of_origin, library, sample_type, drug, rack.concentration) %>%
  summarise(viability = mean(viability, na.rm=T)) %>% 
  ungroup() %>%
  group_by(donor, tissue_of_origin, library, sample_type, drug) %>%
  arrange(as.numeric(rack.concentration)) %>% mutate(x=1:n()) %>%
  summarise(AUC = pracma::trapz(x, viability)) %>% 
  ungroup()
```

Next I visualize tubulin inhibitor phenotypes based on the AUC values.

```{r, results='hide', warning=F, message=F,eval=F}
## tubulin, mek inhibitors
tubi <- c('Paclitaxel', 'Docetaxel', 'Vinblastin')
ctrls <- c('Bortezomib')
meki <- c('Trametinib', 'Binimetinib')

## data frame of tubulin inhibitors
tubulin <- aucs %>% filter(drug %in% c(tubi, ctrls, meki)) %>%
  filter(tissue_of_origin == 'colorectal', sample_type == 'tumour') %>%
  mutate(AUC_resc = plotrix::rescale(AUC, c(0,1)))

## annotation, in manuscript or not?
in_manuscript <- paste0('D0', c('07', '10', '18', '54', '55', '04', 
                                '46', '19', '30', '21', '13', '20', '22', '27'))

anno <- tubulin %>% distinct(donor) %>% 
  mutate(group = ifelse(donor %in% in_manuscript, 'manuscript', 'other')) %>% 
  as.data.frame() %>% column_to_rownames('donor')

## visualize drug effects as heatmap (centered), separate libraries
tubulin %>% filter(library == '08') %>%
  group_by(drug) %>% mutate(AUC = AUC - median(AUC)) %>% ungroup() %>% 
  acast(drug ~ donor, value.var = 'AUC_resc') %>% 
  pheatmap(annotation_col = anno, border_color = '#ffffff', main = 'library 08')

tubulin %>% filter(library == '11') %>%
  group_by(drug) %>% mutate(AUC = AUC - median(AUC)) %>% ungroup() %>% 
  acast(drug ~ donor, value.var = 'AUC_resc') %>% 
  pheatmap(annotation_col = anno, border_color = '#ffffff', main = 'library 11')
```

We see that there are several PDO lines, including newly sreened lines, (in total ~30%) that respond well to tubulin inhibition. As soon as there are expression data available for the new lines this should give us good power to detect associated expression signatures.

```{r, results='hide', warning=F, message=F,eval=F}
## drug clustering
aucs %>% acast(donor ~ drug, value.var = 'AUC') %>% 
  .[,apply(., 2, function(x) !any(is.na(x)))] %>% cor() %>% 
  pheatmap()
```

## Tubulin inhibitor phenotypes in public drug screens

Tublin inhibitors were also used in drug screens at the Sanger institute (1000 cell lines, pancancer). I want to check whether the colorectal cancer cell lines also responded in part to tubulin inhibition. I first parse the data from file (data retrieved from Iorio et al. Cell paper supplementary information).

```{r, results='hide', warning=F, message=F,eval=F}
## parse AUC for tubulin inhibitors
sanger_tubulin <- read.xlsx('../data/cancerxgene/drug_response.xlsx', 
                            sheet = 2, startRow = 6) %>%
  .[,colnames(.) %in% c('X2', 'Paclitaxel', 'Docetaxel', 'Vinblastine', 'Vinorelbine')] %>%
  as_tibble() %>% dplyr::select(cellline = X2, everything())

## select colorectal cancer cell lines
crc_lines <- read.xlsx('../data/cancerxgene/drug_response.xlsx', sheet = 5, startRow = 9) %>% 
  as_tibble() %>% filter(grepl('LARGE_INTESTINE', CCLE.name)) %>% 
  pull(GDSC1000.name)

## plot response for tubulin inhibitors in CRC cell lines
sanger_tubulin_crc <- sanger_tubulin %>% filter(cellline %in% crc_lines) %>% 
  gather(drug, auc, -cellline) %>% group_by(drug) %>% 
  mutate(auc_ct = auc - median(auc, na.rm=T)) %>% ungroup() 

## plot as heatmap
sanger_tubulin_crc %>% acast(drug ~ cellline, value.var = 'auc_ct')  %>% pheatmap()

## bar plot
sanger_tubulin_crc %>% ggplot(aes(cellline, auc)) + 
  geom_bar(stat='identity') + facet_wrap(~drug, ncol=1) + 
  theme(axis.text.x  = element_text(angle=45, hjust=1))
```

Differences can be observed here, too. Interestingly all of the cell lines that we use in the lab (HT29, RKO, HCT116, SW48) seem to not like the Tubulin inhibitors. Where any molecular traits associated with tubulin inhibition? (drug-genotype interactions)

```{r, results='hide', warning=F, message=F,eval=F}
## parse
dgi <- read.xlsx('../data/cancerxgene/drug_response.xlsx', sheet=3, startRow = 38) %>%
  as_tibble()

## filter colorectal/pancancer for tubulin
tubulin_cfe <- dgi %>% as_tibble() %>% 
  filter(Drug.name %in% c('Paclitaxel', 'Docetaxwl', 'Vinblastine', 'Vinorelbine'), 
         Domain %in% c('PANCAN', 'COAD/READ'))
```

There isn't anything really. I can check if there is anything at the gene expression level (above only include mutations, copy nubmer changes and fusions).

```{r, results='hide', warning=F, message=F,eval=F}
## load ready-to-use CCLE gene expression data
load('~/crispr/regulatory_circuits/data/expression/CCLE/CCLE_18q4_TPM_long_all.rda')

## get colorectal lines that overlap with sanger dataset
crc_lines_clean <- tibble(sanger = crc_lines, 
                          clean = gsub('-', '', toupper(sanger)))

## filter lines
crc_expr <- ccle_expr_all %>% filter(cellline %in% crc_lines_clean$clean)

## annotate auc
auc_expr_df <- sanger_tubulin_crc %>% dplyr::select(sanger=cellline, drug, auc) %>% 
  inner_join(crc_lines_clean) %>% distinct() %>% 
  dplyr::select(cellline = clean, everything()) %>% 
  inner_join(crc_expr)
```

I can now use a regression model that estimates correlations between AUC and the expression of a certain gene accouting for differences between individual tubulin inhibitors.

```{r, results='hide', warning=F, message=F,eval=F}
## test correlation for tubulin gene (there might not be anything)
auc_expr_df %>% filter(symbol == 'TUBB') %>% 
  ggplot(aes(auc, tpm)) + geom_point(aes(colour = drug)) + 
  geom_smooth(method='lm')

## same but drugs in separate plots
auc_expr_df %>% filter(symbol == 'TUBB') %>% 
  ggplot(aes(auc, tpm)) + geom_point() + 
  geom_smooth(method='lm') + facet_wrap(~drug)

## evaluate using linear model
summary(lm(auc ~ drug + tpm, data = filter(auc_expr_df, symbol == 'TUBB')))
```

There wasn't any effect for TUBB. I do this now systematically for all genes.

```{r, results='hide', warning=F, message=F,eval=F}
## run linear model for each gene
auc_expr_cor <- auc_expr_df %>% group_by(symbol) %>% 
  group_modify(~ broom::tidy(lm(auc ~ drug + tpm, data = .x))) %>% 
  ungroup() %>% arrange(p.value)
```

What are the top hits? Can we visualize them as scatter plots?

```{r,eval=F}
auc_expr_cor %>% filter(term == 'tpm') 
```

# New microarray data

We did some additional microarray experiments for new PDO lines. I'll analyze these data to look at expression differences between lines that are sensitive to tubulin inhibitors / taxanes and MEK inhibitors.

```{r, results='hide', warning=F, message=F}
## list cel files for analysis
files1 <- list.files('../data/expression/PROMISE/', recursive = T,
                 full.names=T, pattern = '*CEL') %>%
  .[!grepl('HuGene', .)] %>%
  .[!grepl('GSK3|DMSO', .)]

## sample info
sinfo <- tibble(path = files1)
## no duplicate filenames
sinfo <- sinfo %>% mutate(filename2 = gsub('.CEL$', '', basename(path))) %>% 
  arrange(filename2) %>% group_by(filename2) %>% 
  mutate(filename = paste0(filename2, '_', 1:n(), '.CEL')) %>% ungroup() %>%
  dplyr::select(path, filename)

## exclude old versions of 20/21
old2021 <- sinfo %>% filter(grepl('P020T|P021T', path)) %>% 
  filter(grepl('batch1', path))
sinfo <- sinfo %>% anti_join(old2021)

## read the cel files
orgas_bl <- ReadAffy(filenames = sinfo$path, sampleNames = sinfo$filename)

## normalize
expr <- expresso(orgas_bl,
                 bgcorrect.method='rma', 
                 normalize.method='quantiles',
                 pmcorrect.method='pmonly',
                 summary.method='avgdiff')

## extract expr matrix
expr <- exprs(expr)

## probe id to gene symbol
load('../data/meta/affy_probe_mapping_u133plus2.rda')
```

## Sample annotation

### Technical metadata

I read the technical metadata that I received from Johannes and select the important bits for DGE analysis.

```{r, results='hide', warning=F, message=F}
## list of samples and annotation
sample_info <- read.xlsx('../data/meta/Microarray_RNASeq_Metadata_BR.xlsx', 
          detectDates = T) %>% as_tibble()
## remove bad 20 and 21 chips
sample_info <- sample_info %>% group_by(sample_ID_submitted) %>%
  arrange(desc(date_array)) %>% dplyr::slice(1) %>% ungroup()

## merge this info to expr data
chip_files <- tibble(cel_name = colnames(expr)) %>% 
  separate(cel_name, c('sample_ID_submitted', 'chip_info'), 
           sep='_\\(', remove=F)
```

Do we have data for all samples in the Excel file?

```{r, results='hide', warning=F, message=F}
## which are missing?
missing_microarrays <- chip_files %>% 
  right_join(sample_info %>% 
               dplyr::select(PDO, sample_ID_submitted, date_array, 
                             passage, operator.RNA.isolation)) %>% 
  filter(is.na(cel_name)) %>%
  filter(!grepl('GSK|DMSO', sample_ID_submitted))

## get full annotation
sample_info %>% inner_join(missing_microarrays) %>% 
  filter(downstream_analysis != 'RNASeq') %>% 
  write.xlsx('../data/meta/missing_microarrays.xlsx')
```

It seems that everything except for the new batch and arrays with IDs 46 and 47 is there. I combine chip-files and sample annotation for downstream analysis.

```{r, results='hide', warning=F, message=F}
sample_info <- chip_files %>% 
  inner_join(sample_info %>% 
               dplyr::select(PDO, sample_ID_submitted, date_array, passage,
                             operator.RNA.isolation)) 
```

### Clinical metadata

I parse clinical metadata from the Excel file that Johannes sent to me.

```{r, results='hide', warning=F, message=F}
## parse file
clinical_meta <- read.xlsx('../data/meta/ClinicalData_Pat_m_Microarray.xlsx') %>% tbl_df()

## select relevant info
clinical_meta <- clinical_meta %>% 
  dplyr::select(Donor_ID, Type, `Biopsy_primary-metastasis`, age, 
                gender, PDO_propagation, Histology, Location, `Stage.(biopsy)`)
```

I add the clinical metadata to the rest of the metadata.

```{r, results='hide', warning=F, message=F}
sample_info <- sample_info %>% mutate(Donor_ID = substr(PDO, 1, 4)) %>% 
  inner_join(clinical_meta %>% mutate(Donor_ID = stringr::str_trim(Donor_ID))) %>%
  mutate(Location = ifelse(Location == 'Sigm', 'Sigma', Location))
```

This looks good, only organoid 93 is missing.

## Quality control

I will do PCA/clustering to select the good samples and identify potential batches.

```{r, results='hide', warning=F, message=F}
## boxplot to check normalization
boxplot(log(expr)[sample(1:nrow(expr), 2000),])
```

I can colour code all kinds of meta-variables here in order to see what influences the PCA plot the most.

```{r, results='hide', warning=F, message=F}
## pca
pca_out <- prcomp(t(log(expr)))
## plot results
as_tibble(pca_out$x, rownames='cel_name') %>% 
  left_join(sample_info) %>%
  ggplot(aes(PC1, PC2, label = PDO,
             colour = Location, 
             shape = gender,
             size = date_array)) + 
  geom_point() + 
  ggrepel::geom_text_repel()

## to check pca loadings
# pca_out$rotation %>% as_tibble(rownames='probe') %>% 
#   inner_join(id_mapping_u133 %>% distinct(probe, symbol) %>% 
#                filter(!is.na(probe))) %>% arrange(PC1) %>% 
#   dplyr::select(probe, symbol, PC1:PC5)
```

It's clear that biological variation dominates the PCA plot, PC1 seems to separate colorectal and gastric/esophageal cancers. I would say that the operator likely doesn't make a difference. Also there do not seem to be any obvious date-related batch effects. I plot the genes that dominate the PC loadings.

```{r, results='hide', warning=F, message=F}
## expr, long data frame with annotation
expr_long <- expr %>% as_tibble(rownames = 'probe') %>% 
  gather(cel_name, expr, -probe) %>% 
  left_join(sample_info) %>%
  left_join(id_mapping_u133 %>% distinct(probe, symbol) %>% 
              filter(!is.na(probe))) %>%
  filter(cel_name %in% sample_info$cel_name)

## plot some genes that dominate the loadings against some metadata
poi <- as_tibble(pca_out$rotation, rownames='probe') %>% 
  left_join(id_mapping_u133 %>% distinct(probe, symbol) %>% 
              filter(!is.na(probe))) %>%
  arrange(PC2) %>% pull(symbol) %>% unique() %>% .[!is.na(.)] %>% head(50)

expr_long %>% filter(probe == '206199_at') %>% 
  ggplot(aes(date_array, log(expr), colour = as.factor(`Stage.(biopsy)`))) + 
  geom_point()
```

I use regression analysis to look into some of the PC loadings to figure out what variables they represent.

```{r, results='hide', warning=F, message=F}
## plot most expressed probes only
df <- expr_long %>% filter(probe == '227174_at')

summary(lm(log(expr) ~ date_array + passage + Location + `Stage.(biopsy)`, data = df))
```

It's not 100% clear but it seems that if anything sub-location of the colorectal tumors (rectum vs descendum) seems to make a difference.

## Differential gene expression

### Tubulin inhibitor sensitive lines

To test differential gene expression analyses I look at differences between Tubulin inhibitor-sensitive vs resistant organoid lines. First I annotate (derrived from CTG heatmaps), which lines are resistant and which are senstivie to tubulin inhibitors

```{r, results='hide', warning=F, message=F}
## lists of organoid sets, so that I can split old and newly screened ones
old_pdo <- c('D018', 'D007', 'D010', 'D046', 'D015', 'D027', 
             'D052', 'D004', 'D054', 'D021', 'D030', 'D020', 
             'D022', 'D013', 'D019', 'D055')

new_pdo <- c('D036', 'D104', 'D054', 'D086', 'D073', 'D089', 'D105', 'D115',
             'D082', 'D090', 'D080', 'D055', 'D098')

## tubulin inhibitor sensitive lines
tub_sens <- c('D020', 'D022', 'D073', 'D105', 'D115')
```

I now add this to the sample info file, select only colorectal cancer chips, and test whether I can identify differentially expressed genes that could explain.

```{r, results='hide', warning=F, message=F}
## sample info for taxan response, only CRC no line 93
taxan_sample_info <- sample_info %>% 
  filter(Type == 'CRC', !is.na(Type),
         Donor_ID %in% c(old_pdo)) %>% # incase i want to seaprate old and new
  mutate(taxan_response = ifelse(Donor_ID %in% tub_sens, 
                                 'sensitive', 'resistant')) %>% 
  filter(!is.na(taxan_response)) %>% 
  arrange(cel_name)

## make an expresion matrix (log-scale), select samples in taxan sample info
taxan_expr <- log(expr[,colnames(expr) %in% taxan_sample_info$cel_name])

## make sure that sample info and expr matrix match
identical(colnames(taxan_expr), taxan_sample_info$cel_name)

## model matrix, account for date, sex, tumor stage, location
mm <- model.matrix(~ date_array + gender + `Stage.(biopsy)` + Location + taxan_response, 
                   data = taxan_sample_info)
## run DGE
fit <- lmFit(log(taxan_expr), mm)
fit <- eBayes(fit)

## results, annotate gene symbol, look at diffrent coefficients to adjust covariates
taxan_res <- topTable(fit, coef = 8, n = Inf) %>% 
  as_tibble(rownames = 'probe') %>% 
  left_join(id_mapping_u133) %>% distinct()
```

I make a few plots to look at genes that change significantly by date or location.

```{r, results='hide', warning=F, message=F}
## ganab changes with date
expr_long %>% filter(probe == '201183_s_at') %>% 
    ggplot(aes(date_array, log(expr), colour = Location)) + 
    geom_point()

## XIST is location (desc) dependent
expr_long %>% filter(probe == '224590_at') %>% 
  ggplot(aes(gender, log(expr), label = PDO)) + 
  geom_jitter(width=0.2) + 
  stat_summary(fun.ymin = 'median', fun.ymax = 'median', fun.y = 'median', 
               color = 'red', geom='crossbar', width = 0.5)

## TNFSF11 is location (desc) dependent
expr_long %>% filter(probe == '210643_at') %>% 
    ggplot(aes(Location, log(expr))) + 
    geom_point() + 
  stat_summary(fun.ymin = 'mean', fun.ymax = 'mean', fun.y = 'mean', 
               color = 'red', geom='crossbar', width = 0.5) + 
  theme(axis.text.x = element_text(angle=45, hjust=1))

## TNFSF11 is location (desc) dependent
expr_long %>% filter(probe == '207981_s_at', symbol == 'ESRRG') %>% 
  filter(Type == 'CRC', !is.na(Type), Donor_ID %in% c(old_pdo, new_pdo)) %>%
  mutate(taxan_response = ifelse(Donor_ID %in% tub_sens, 
                                 'sensitive', 'resistant')) %>%
    ggplot(aes(taxan_response, log(expr), colour = gender)) + 
    geom_jitter(width = 0.2) + 
  stat_summary(fun.ymin = 'mean', fun.ymax = 'mean', fun.y = 'mean', 
               color = 'red', geom='crossbar', width = 0.5) + 
  theme(axis.text.x = element_text(angle=45, hjust=1))
```

These make sense clearly. Also the sex specific gene expression makes a lot of sense and is dominated by XIST. 

Now what are the differentially expressed genes between Tubulin sensitive and insenstive lines?

```{r, results='hide', warning=F, message=F}
## APC old organoids
expr_long %>% filter(symbol == 'APC') %>% 
  filter(Type == 'CRC', !is.na(Type), Donor_ID %in% old_pdo) %>%
  mutate(taxan_response = ifelse(Donor_ID %in% tub_sens, 
                                 'sensitive', 'resistant')) %>%
  ggplot(aes(taxan_response, log(expr), colour = gender)) + 
  geom_jitter(width = 0.2) + 
  stat_summary(fun.ymin = 'mean', fun.ymax = 'mean', fun.y = 'mean', 
               color = 'red', geom='crossbar', width = 0.5) + 
  facet_wrap(~ probe)

## AXIN2
expr_long %>% filter(symbol == 'AXIN2') %>% 
  filter(Type == 'CRC', !is.na(Type), Donor_ID %in% old_pdo) %>%
  mutate(taxan_response = ifelse(Donor_ID %in% tub_sens, 
                                 'sensitive', 'resistant')) %>%
  ggplot(aes(taxan_response, log(expr), colour = gender)) + 
  geom_jitter(width = 0.2) + 
  stat_summary(fun.ymin = 'mean', fun.ymax = 'mean', fun.y = 'mean', 
               color = 'red', geom='crossbar', width = 0.5) + 
  facet_wrap(~ probe)
```

## Morphology clusters

Niklas clustered PDOs from different lines according to their morphological phenotype (single organoid measurements extracted from CNN segmented images). I want to explore if linear models can be used to map these clusters to gene expression signatures.
First, however, I perform clustering analyses to see whether there are interesting clusters that are observable by eye.

I start by loading the proportions of healthy organoids for each line that fall into each cluster (provided by Niklas).

```{r, results='hide', warning=F, message=F}
## parse cluster proportions from file.
cluster_props <- read_csv('../data/meta/cluster_proportions.csv') %>% 
  gather(cluster, prop, -line)

## list of clusters with high variance (knee plot)
cluster_props %>% ggplot(aes(cluster, prop, fill = line)) + 
  geom_bar(stat='identity', position = 'dodge') + 
  facet_wrap(~cluster, scales = 'free_x')
```

### Cluster M6

Let's look at cluster 6 - the cluster with the biggest variance. I first plot a heatmap of variable genes, highlighting cluster 6 proportions for each line. I'm wondering whether cluster 6 drives the overall variance in gene expression.

```{r, results='hide', warning=F, message=F}
## downsized gene expression dataset
expr_fil <- expr_long %>% filter(PDO %in% cluster_props$line)

## main probe for each symbol (max expr.)
main_probes <- expr_fil %>% 
  group_by(symbol, probe) %>% 
  summarise(max_expr = max(expr), mean_expr = mean(expr)) %>% ungroup() %>% 
  group_by(symbol) %>% arrange(desc(mean_expr), desc(max_expr)) %>%
  dplyr::slice(1) %>% ungroup() %>%
  ## some probes match to several symbols - just take one of them
  group_by(probe) %>% dplyr::slice(1) %>% ungroup()

## add entrez ID
egids <- bitr(main_probes$symbol, 
              fromType = 'SYMBOL', 
              toType = 'ENTREZID', 
              OrgDb = 'org.Hs.eg.db', drop=F) %>%
  as_tibble() %>% `colnames<-`(c('symbol', 'entrez'))
main_probes <- main_probes %>% left_join(egids) %>% 
  group_by(symbol) %>% dplyr::slice(1) %>% ungroup()

## cluster 6 annotation for heatmap
anno <- cluster_props %>% filter(cluster == 'm4') %>%
  left_join(
    expr_fil %>% dplyr::select(cel_name, line=PDO) %>% distinct()
  ) %>%
  drop_na() %>% 
  dplyr::select(cel_name, m4_prop = prop) %>%
  as.data.frame() %>% column_to_rownames('cel_name')

## plot gene expression
expr_mat <- expr_fil %>%
  filter(cel_name %in% rownames(anno)) %>% 
  inner_join(main_probes %>% distinct(probe, symbol)) %>%
  mutate(log_expr = log(expr)) %>% 
  acast(symbol ~ cel_name, value.var = 'log_expr')

## draw heatmap
expr_mat[order(apply(expr_mat, 1, sd), decreasing=T)[1:50],] %>% 
  pheatmap(annotation_col = anno)
```

It does seem that cluster 6 proportions are to an extent correlated with overall gene expression patterns. However, ther is organoid 4 that clusters with high M6 lines but that does not have many organoids in M6. I next aim to use a linear model to find expression markers for M6.

```{r, results='hide', warning=F, message=F,eval=F}
cluster_dge <- map(c('m4'), function(cl){
  ## sample info
  si <- cluster_props %>% filter(cluster == cl) %>%
    left_join(
      expr_fil %>% dplyr::select(cel_name, line=PDO) %>% distinct()
    ) %>% drop_na() %>% inner_join(sample_info) %>%
    arrange(cel_name)
  
  ## model matrix
  mm <- model.matrix(~ gender + prop, data = si)
  
  ## expression matrix (all probes)
  expr_mat <- expr_fil %>% filter(cel_name %in% si$cel_name) %>% 
    inner_join(distinct(main_probes, probe, entrez)) %>%
    mutate(log_expr = log(expr)) %>%
    distinct(cel_name, entrez, log_expr) %>% drop_na() %>%
    acast(entrez ~ cel_name, value.var = 'log_expr')
  
  ## do sample-info and expr. matrix match?
  identical(colnames(expr_mat), si$cel_name)
  
  ## fit linear model
  fit <- lmFit(expr_mat, mm)
  fit <- eBayes(fit)
  
  ## go analysis
  go <- topGO(goana(fit, coef = 3), ontology = 'BP') %>% 
    as_tibble(rownames = 'goid')
  
  ## results
  dge_res <- topTable(fit, n=Inf, coef = 3) %>% 
    as_tibble(rownames='entrez') %>% 
    left_join(distinct(main_probes, entrez, symbol))
  
  ## filtered for interesting probes
  return(list(dge = dge_res, go = go))
})
```

## Morphologically different lines

Based on the above results I think that it makes much more sense to compare individual lines based on where most of their WT organoids are.

### Covariates

I first have to perform DGE for a number of confounding factors such as sex and TP53 mutation status. I can then flag these hits in the comparison of individual lines.

```{r, results='hide', warning=F, message=F}
## sample info
si <- sample_info %>% filter(PDO %in% cluster_props$line) %>%
  filter(PDO != 'D021T01') %>%
  arrange(cel_name)

## annotate TP53 mutation status
si <- si %>% 
  mutate(tp53 = ifelse(PDO %in% c('D013T01', 'D019T01', 'D020T01', 'D020T01', 'D022T01', 'D027T01'),
                       'mut', 'wt'),
         igf2 = ifelse(PDO %in% c('D055T01', 'D054T01', 'D046T01', 'D018T01', 'D004T01'), 'igf2', 'wt'))

## expression matrix (all probes)
expr_mat <- expr_fil %>% filter(cel_name %in% si$cel_name) %>% 
  inner_join(distinct(main_probes, probe, entrez)) %>%
  mutate(log_expr = log(expr)) %>%
  distinct(cel_name, entrez, log_expr) %>% drop_na() %>%
  acast(entrez ~ cel_name, value.var = 'log_expr')

## model matrix
mm <- model.matrix(~ tp53 + gender + igf2, data = si)

## do sample-info and expr. matrix match?
identical(colnames(expr_mat), si$cel_name)

## fit linear model
fit <- lmFit(expr_mat, mm)
## define contrasts (lines to compare)
fit <- eBayes(fit)

## p53 genes
p53_res <- topTable(fit, n=Inf, coef = 'tp53wt') %>% 
  as_tibble(rownames='entrez') %>% 
  left_join(distinct(main_probes, entrez, symbol)) %>%
  filter(adj.P.Val < 0.01) %>% pull(symbol)

## sex genes
sex_res <- topTable(fit, n=Inf, coef = 'genderm') %>% 
  as_tibble(rownames='entrez') %>% 
  left_join(distinct(main_probes, entrez, symbol)) %>%
  filter(adj.P.Val < 0.05) %>% pull(symbol)

## igf2 genes
igf2_res <- topTable(fit, n=Inf, coef = 'igf2wt') %>% 
  as_tibble(rownames='entrez') %>% 
  left_join(distinct(main_probes, entrez, symbol)) %>%
  filter(adj.P.Val < 0.05) %>% pull(symbol)
```

### Differentially expressed genes

I fit a big model including all organoid lines and then compare individual lines via contrasts.

```{r, results='hide', warning=F, message=F}
## model matrix
mm <- model.matrix(~ 0 + PDO, data = si)

## do sample-info and expr. matrix match?
identical(colnames(expr_mat), si$cel_name)

## fit linear model
fit <- lmFit(expr_mat, mm)
## define contrasts (lines to compare)
cont <- makeContrasts(c1='(PDOD055T01+PDOD046T01+PDOD054T01)/3-(PDOD022T01+PDOD007T01+PDOD027T01)/3', 
                      levels = mm)
fit_cont <- contrasts.fit(fit, cont)
fit_cont <- eBayes(fit_cont)

## gene set enrichemnt  
go <- topGO(goana(fit_cont, coef = 'c1'), ontology = 'BP') %>% 
  as_tibble(rownames = 'goid')

## results
top3_res <- topTable(fit_cont, n=Inf, coef = 'c1') %>% 
  as_tibble(rownames='entrez') %>% 
  left_join(distinct(main_probes, entrez, symbol)) %>%
  mutate(tp53 = ifelse(symbol %in% p53_res, T, F),
         sex = ifelse(symbol %in% sex_res, T, F),
         igf2 = ifelse(symbol %in% igf2_res, T, F))
```

Plot to compare genes between lines.

```{r, results='hide', warning=F, message=F}
expr_long %>% 
  filter(symbol == 'WSB2', 
         PDO %in% c('D055T01', 'D046T01', 'D054T01', 'D022T01', 'D007T01', 'D027T01')) %>% 
  ggplot(aes(PDO, log(expr))) + 
  geom_point() + facet_wrap(~probe)
```

I visualize differentially expressed genes implicated in MTOR signaling in a volcano plot.

```{r, results='hide', warning=F, message=F}
volc_data <- top3_res %>% 
  mutate(group = ifelse(symbol %in% c('IGF2', 'MIR483', 'H19', 'INS-IGF2'), 'igf2', 'none')) %>%
  mutate(label = ifelse((group != 'none') & (adj.P.Val < 0.01), symbol, ''))

co05 <- volc_data %>% filter(adj.P.Val > 0.05) %>% pull(P.Value) %>%
  head(1) %>% log10() %>% `*`(-1)

ggplot() +
  ggrastr::geom_point_rast(data=subset(volc_data, group == 'none'),
             aes(logFC, -log10(P.Value)), colour='#dddddd') +
  geom_point(data=subset(volc_data, group == 'igf2'),
             aes(logFC, -log10(P.Value)), colour='#4285f4') +
  geom_hline(yintercept = co05, linetype='dashed') +
  ggrepel::geom_text_repel(data=subset(volc_data, group == 'igf2'),
                  aes(logFC, -log10(P.Value), label=label)) +
  ylab('P-value [-log10]') + xlab('Fold change [log2]') + 
  scale_y_continuous(expand=c(0,0))
```

### Gene set enrichment

Gene set enrichment analysis

```{r, results='hide', warning=F, message=F}
## gene ranks for enrichment, best T-value for each gene
ranks <- top3_res %>%
  filter(!tp53, !sex) %>%
  arrange(desc(abs(t)))
ranks <- setNames(ranks$t, ranks$entrez)

## get pathway annotation
pathways <- fgsea::reactomePathways(names(ranks))

## perform GSEA
enr <- fgsea::fgsea(pathways, ranks, nperm = 1e5, maxSize=500)
enr <- as_tibble(enr) %>% arrange(pval)

## genes for interesting terms
eg <- enr %>% filter(pathway == 'IGF1R signaling cascade') %>% pull(leadingEdge) %>% unlist()


## a function to draw barcode plots
custom_barcode_plot <- function(stat_vector, sig_genes, term){
  ## generate barcode plot
  bc_plot <- plotEnrichment(sig_genes, stat_vector)
  
  ## remove unwanted layers
  bc_plot$layers <- list()
  
  ## add barcode at the bottom
  lowest_pos <- min(bc_plot$data[,2])
  dash_length <- abs(purrr::reduce(range(bc_plot$data[,2]), `-`)*0.1)
  middle <- which.min(abs(sort(stat_vector, decreasing=T)))
  
  bc_plot_custom <- bc_plot + geom_segment(aes(x=x, xend=x), y=lowest_pos,
                           yend=lowest_pos-dash_length) + 
    geom_line(colour='#4daf4a') + 
    geom_hline(yintercept=lowest_pos, colour='#cccccc') + 
    geom_hline(yintercept=0, colour='#cccccc') + xlab('') +
    theme_classic() +
    geom_tile(data=tibble(rank=1:length(stat_vector), 
                          y=lowest_pos-(1.25*dash_length)), 
              aes(x=rank, y=y, fill=rank),
                  width=1,
                  height=0.5*dash_length) +
    scale_fill_gradient2(low ='#b2182b', high='#2166ac', 
                         mid='#f7f7f7', midpoint = middle) + 
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(expand = c(0, 0)) +
    theme(panel.grid=element_blank(), 
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          legend.position = 'none') + 
    ggtitle(term) +
    ylab('Enrichment score')
  
  return(bc_plot_custom)
}

signature_genes <- pathways['IGF1R signaling cascade'] %>% unlist()

custom_barcode_plot(ranks, signature_genes,
                    'IGF1R signaling')
```


# Session info

```{r}
sessionInfo()
```