---
title: "Differential expression analyses of the PROMISE project"
author: "Benedikt Rauscher"
date: "9/8/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, echo=F, message=F, warning=F, results='hide'}
library(tidyverse)
library(limma)
library(openxlsx)
```

# Loading data

Data are saved as simple tab-separated file and can be loaded into R.

```{r, results='hide', warning=F, message=F}
expr <- read_tsv('~/tumor_analysis/molecular_subtypes/data/expression/organoids/PROMISE_gene_expression.tsv')
```

# Annotation of samples

## Morphology

We noticed that organoid lines show distinct morphological differences. It might be interesting to check whether there are differentially expressed genes between morphologically distinct groups that could explain these phenotypes and shed more light on the tumour characteristics of these lines.

```{r, results='hide', warning=F, message=F}
## load file with morphological group annotation
morph_anno <- read.xlsx('rsync_data/170908_morphology_organoids_classification.xlsx') %>% tbl_df
morph_anno <- morph_anno %>% dplyr::rename(line=Line) %>% 
  mutate(line = ifelse(grepl('00', line), gsub('00', '0', line), line))
```

# Differential expression analysis

## Differences in morphology

First we are interested whether differences in cell morphology can be explained by gene expression patterns. We perform a number of differential expression analyses to investigate this.

### Solid versus cystic organoids

Organoid lines can be visually distinguished into solid and cystic lines.

```{r, results='hide', warning=F, message=F}
## expression matrix for limma
expr_mat <- expr %>% unite(sample, line, rep, sep='_') %>% 
  dplyr::select(symbol, sample, expr) %>% distinct() %>% 
  spread(sample, expr) %>% data.frame() %>% 
  `rownames<-`(.$symbol) %>% dplyr::select(-symbol) %>%
  .[,] %>% as.matrix()

## annotation frame
anno <- tibble(sample=colnames(expr_mat)) %>% 
  separate(sample, c('line', 'rep'), sep='_', remove=F) %>% 
  left_join(morph_anno) %>% 
  filter(morphology_class != 3) %>% 
  mutate(mc = as.factor(ifelse(morphology_class %in% 1:2, 1, 2)))

## perform DGE analysis
mm <- model.matrix(~mc, data=anno)
fit <- lmFit(expr_mat[,colnames(expr_mat) %in% anno$sample], design=mm)
efit <- eBayes(fit)
```