---
title: "tgfb_dge"
author: "Niklas Rindtorff"
date: "4/10/2018"
output: 
   BiocStyle::html_document2:
    code_folding: hide
    df_print: paged
    toc_depth: 3
    toc_float: true
    fig_height: 6
    fig_width: 9.5
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
#Intro
##Load packages 
```{r, message=FALSE, warning=FALSE}
library(tidyverse)
library(readxl)
library(patchwork)
library(fgsea)
```

##Load data
```{r, message=FALSE, warning=FALSE}
dge_52<- read_excel("~/PowerFolders/2017-08 PDO [Johannes Betge Niklas Rintdorff]/Tables/Expression_Morphology/PROMISE_differences_morphology_classes.xlsx", 
    sheet = "class5 - class2 DGE") %>% mutate(log_p = -log10(P.Value)) 
enrichment_52 <- read_excel("~/PowerFolders/2017-08 PDO [Johannes Betge Niklas Rintdorff]/Tables/Expression_Morphology/PROMISE_differences_morphology_classes.xlsx", 
    sheet = "class5 - class2 Enrichment")
```

#Vulcano plot
I search for biological processes which are enriched in compact organoids when compared to cystic phenotypes.
```{r, message=FALSE, warning=FALSE}
enrichment_52 %>% filter(toi == TRUE, NES < 0) %>% dplyr::select(term, NES)
```

I now plot a vulcano plot showing the most significantly de-regulated genes. On the left side are genes that upregulated in dense organoids. 

Brown are genes belonging to the term: Signaling by TGF-beta Receptor Complex
Blue are genes belonging to the term: Transcriptional activity of SMAD2/SMAD3:SMAD4 heterotrimer

```{r, message=FALSE, warning=FALSE}
#
toif <- "Signaling by TGF-beta Receptor Complex"
aa <- enrichment_52 %>% 
filter(term %in% toif) %>% #lipid metabolism
#arrange(desc(Count)) %>%
#arrange((Count)) %>%
.$leadingEdge %>%
  stringr::str_split(pattern = ", ") %>%
  unlist%>%
  as.numeric()

#
toif <- "Regulation of TP53 Activity"
toif <- "Transcriptional activity of SMAD2/SMAD3:SMAD4 heterotrimer"
bb <- enrichment_52 %>% 
filter(term %in% toif) %>% #lipid metabolism
#arrange(desc(Count)) %>%
#arrange((Count)) %>%
.$leadingEdge %>%
  stringr::str_split(pattern = ", ") %>%
  unlist%>%
  as.numeric()

 #t100_aw <- c("TGFBR1", "TGFBR2", "SMAD4")

t100_aw <- dge_52 %>% filter(log_p > 4 & entrez %in% c(aa, bb, "TGFB1")) %>% .$symbol %>% unique()

#Main Figure
fda <- dge_52 %>% select(log_p, adj.P.Val) %>% mutate_all(funs(round(., 4))) %>% filter(adj.P.Val == 0.0500) %>% .$log_p %>% mean()

ggdf <- dge_52 %>% 
  mutate(goi = ifelse(entrez %in% aa, "A", FALSE)) %>% 
  mutate(goi = ifelse(entrez %in% bb, "B", goi)) %>%
  #mutate(goi = ifelse(entrez %in% cc, "C", goi)) %>%
  #mutate(goi = ifelse(entrez %in% dd, "D", goi)) %>% 
  mutate(goi_label = ifelse(symbol %in% t100_aw, TRUE, FALSE))



ggdf %>%
  #mutate(alpha_val = if_else(entrez %in% c(ng), 1, 1)) %>%
  ggplot(aes(logFC, log_p)) + 
  geom_point(data = ggdf %>% filter(goi == FALSE), color = "grey") + 
  #geom_point(data = ggdf %>% filter(goi == "D", logFC < 0), color = "#01665e") +
  geom_point(data = ggdf %>% filter(goi == "A"), color = "#8c510a") + 
  geom_point(data = ggdf %>% filter(goi == "B"), color = "#5ab4ac") +
  #geom_point(data = ggdf %>% filter(goi == "C"), color = "#d8b365") +
  #geom_point(data = ggdf %>% filter(goi == "D", logFC > 0), color = "#01665e") +
  geom_point(data = ggdf %>% filter(goi_label == TRUE), color = "#8c510a") +
  ggrepel::geom_text_repel(data = ggdf %>% filter(goi_label == TRUE & log_p > 2.5), aes(label = symbol)) +
  geom_rug(alpha=.01) +
  #facet_grid(~contrast) + 
  theme_classic() + 
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = fda) + 
  scale_color_manual(values=c("black", "red"))+ ggtitle("Dense vs. Cystic differential gene expression")
```

The genes Serpin1 and Trim33 drive the TGFb phenotype.

I now want to validate the enrichment of TGFb induced genes in the compact phenotype

##GSEA for validation  
Using a published stem cell signature I want to identify differential expression of these target genes. 
First I load the signature dataset and clean it up. 

```{r,message=FALSE, warning=FALSE}
signatures <- lapply(c("ISC genes", "Proliferation genes", "Late TA genes", "Lgr5 genes"), function(t) read_excel("~/mco/local_data/expression/array_annotation/1-s2.0-S193459091100110X-mmc2_bettle.xls", 
    sheet = t, skip = 3) %>% .[,c(1:3)] %>% mutate(signature = t)) %>%
  bind_rows() %>%
  mutate_all(~gsub('\\*', '', .)) %>% 
  filter(!grepl(.$Gene,pattern = "NA")) %>%
  filter(!grepl(.$Gene,pattern = "includes")) %>% #I just kick out genes which include this phrase for now
  split(., .$signature) %>%
  map(~ .$Gene)

signatures <- append(signatures, list.files(path = "~/PowerFolders/2017-08 PDO [Johannes Betge Niklas Rintdorff]/Vignettes/02_phenotype_differences/gsea", full.names = TRUE) %>% 
  map(function(p) read_csv(p, col_names = FALSE, skip = 2)) %>%
  map(~ .$X1) %>%
  `names<-`(c(list.files(path = "~/PowerFolders/2017-08 PDO [Johannes Betge Niklas Rintdorff]/Vignettes/02_phenotype_differences/gsea")) %>% substr(., 1, nchar(.)-4) %>% tolower())
)

names(signatures)
```

Next I implement a function written by B. Rauscher to create pretty GSEA plots. 

```{r, message=FALSE, warning=FALSE}
custom_barcode_plot <- function(df, sig){
  ## named vector of gene-level stats
  stat_vector <- setNames(df$t, df$symbol) #The GSEA will run on mouse genes
  ## genes in signature
  sig_genes <- signatures[[sig]]
  
  ## generate barcode plot
  bc_plot <- plotEnrichment(sig_genes, stat_vector)
  
  ## remove unwanted layers
  bc_plot$layers <- list()
  
  ## add barcode at the bottom
  lowest_pos <- min(bc_plot$data[,2])
  dash_length <- abs(purrr::reduce(range(bc_plot$data[,2]), `-`)*0.1)
  middle <- which.min(abs(sort(df$t, decreasing=T)))
  
  bc_plot_custom <- bc_plot + geom_segment(aes(x=x, xend=x), y=lowest_pos,
                           yend=lowest_pos-dash_length) + 
    geom_line(colour='#4daf4a', size = 1.5) + 
    geom_hline(yintercept=lowest_pos, colour='#cccccc') + 
    geom_hline(yintercept=0, colour='#cccccc') + xlab('') +
    theme_classic() +
    geom_tile(data=tibble(rank=1:length(stat_vector), 
                          y=lowest_pos-(1.25*dash_length)), 
              aes(x=rank, y=y, fill=rank),
                  width=1,
                  height=0.5*dash_length) +
    scale_fill_gradient2(low ='#b2182b', high='#2166ac', 
                         mid='#f7f7f7', midpoint = middle) + 
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(expand = c(0, 0)) +
    theme(panel.grid=element_blank(), 
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          legend.position = 'none') + 
    ggtitle(paste(sig, 'signature')) +
    ylab('Enrichment score')
  
  return(bc_plot_custom)
}
```

I now run multiple GSEAs on the 5-2 contrast

```{r, message=FALSE, warning=FALSE}
gsea_out <- dge_52 %>% 
                     #filter(contrast == c) %>%
                      mutate(ranklist = setNames(.$t, .$symbol)) %>%
                      .$ranklist %>% 
                      fgsea(signatures, ., nperm = 10000)
gsea_out
```

I plot an example barcode plot for the 5-2 contrast

```{r, message=FALSE, warning=FALSE}
dge_52 %>% custom_barcode_plot(., "tgfb_up.v1_up")
```



